src/runtime/HalideBuffer.h

```diff
@@ -710,6 +711,27 @@ class Buffer {
                       "Can't convert from a Buffer with static dimensionality to a Buffer with different static dimensionality");
     }
 
+    template<typename SomeType>
+    inline SomeType align_up(SomeType value, size_t alignment) {
+        // Doing pointer -> integer -> pointer conversions can force
+        // the compiler to pessimize optimizations (see e.g.
+        // https://www.ralfj.de/blog/2020/12/14/provenance.html),
+        // so, if our compiler has builtins for this, use them;
+        // if not, fall back to casting as necessary.
+#if defined(__has_builtin) && __has_builtin(__builtin_align_up)
+        return __builtin_align_up(value, alignment);
+#else
+        if constexpr (std::is_pointer<SomeType>::value) {
```

Comment:
How about this implementation for align_up:
```
ptr += ((-(uintptr_t)(ptr)) & (alignment - 1));
```
Curious if that passes the clang-tidy checker. It just adds a computed offset to a pointer, where the offset depends on casting the pointer to an int. It never casts an int to a ptr.